### ISSUE TITLE
False error report when using Union and Generics in Atom IDE

### ISSUE THREAD

**Issue by JohnAtFenestra** [TypeStrong/atom-typescript#945]
This turned up while following Anders Hejlsberg's [talk on Google](https://www.youtube.com/watch?v=Ut694dsIa8w). The code snippet is as follows and was copied directly from Anders' talk. I extracted this function into a single file named `snippet.ts`:

``` typescript
function sortBy<T>(a: T[], key: string | ((item: T) => any ) ) {
    var result = a.slice(0);
    // The next line produces an IDE error message of
    // An index expression must be of type 'string', 'number', 'symbol', or 'any'.
    // x[key]
    var callback = typeof key === "string" ? x => x[key] : key; 
    result.sort(function(x, y) {
        var kx = callback(x);
        var ky = callback(y);
        return kx > ky ? 1 : kx < ky ? -1 : 0;
    });
    return result;
}
```

![image](https://cloud.githubusercontent.com/assets/619697/15160871/c04acd50-16c9-11e6-8110-e3e05264ded6.png)

However, saving the test file emits the following transpiled code which executes correctly when tested:

``` javascript
function sortBy(a, key) {
    var result = a.slice(0);
    var callback = typeof key === "string" ? function (x) { return x[key]; } : key;
    result.sort(function (x, y) {
        var kx = callback(x);
        var ky = callback(y);
        return kx > ky ? 1 : kx < ky ? -1 : 0;
    });
    return result;
}
```

By comparison, the following simplified function does not produce an error

``` typescript
function myTest<T>(a: T[], key: string | number ) {
    var value = typeof key === "string" ? x => x[key] : key;
}
```

**Update**

While digging deeper into the video, I discovered a workaround from a coding standpoint. If we explicitly tell the checker the type, the error message disappears:

``` typescript
function sortBy<T>(a: T[], key: string | ((item: T) => any ) ) {
    var result = a.slice(0);
    var callback = typeof key === "string" ? x => x[<string>key] : key; 
    result.sort(function(x, y) {
        var kx = callback(x);
        var ky = callback(y);
        return kx > ky ? 1 : kx < ky ? -1 : 0;
    });
    return result;
}
```
